virttest/storage.py

```diff
@@ -656,7 +657,34 @@ def __init__(self, params, root_dir, tag):
             self.snapshot_format = ss_params.get("image_format")
 
         self.image_access = ImageAccessInfo.access_info_define_by_params(
-                                                   self.tag, self.params)
+            self.tag, self.params)
+
+        self.data_file = self.external_data_file_defined_by_params(
+            params, root_dir, tag)
+
+    @classmethod
+    def external_data_file_defined_by_params(cls, params, root_dir, tag):
+        """Link image to an external data file."""
+        enable_data_file = params.get("enable_data_file", "no") == "yes"
+        image_format = params.get("image_format", "qcow2")
+        if not enable_data_file:
+            return
+        elif image_format != "qcow2":
+            raise ValueError("The %s format does not support external "
+                             "data file" % image_format)
+        else:
+            image_size = params["image_size"]
+            base_name = os.path.basename(params["image_name"])
+            data_file_path = params.get("data_file_path",
+                                        os.path.join(root_dir, "images",
+                                                     "%s.data_file"
+                                                     % base_name))
```

Comment:
I didn't see any file creation procedure been invoked in `QemuImg.create`, so would you please explain how it is created? thanks.

Reply:
Please refer to the following codes, hope to help you. Thanks.

QemuImg.__init__(self, params, root_dir, tag)
    storage.QemuImg.__init__(self, params, root_dir, tag)

storage.QemuImg.__init__(self, params, root_dir, tag)
    self.data_file = self.external_data_file_defined_by_params(params, root_dir, tag)

QemuImg._parse_options(self, params)
    if self.data_file:
        options.extend(
            ("data_file=%s" % self.data_file.image_filename,
             "data_file_raw=%s" % params.get("data_file_raw", "off")))

QemuImg.create(self, params, ignore_errors=False)
    ......
    options = self._parse_options(params)
    ......

Reply:
I mean, do we need to create the data file explicitly before the qcow2 image creation? In other words, about the management of data file, what actions are required?

I tested the image creation with a non-existent data file, and it turns out that the data file would be created when the `create` command being executed, that means we do can let the data file to be created when it is being used at the first time, but I'm not sure if it is correct for all the usecases. And, for image deletion it seems missing the action for data files, because obviously tools other than `qemu-img` won't remove data files explicitly.

Reply:
for external data file, it's a image file contains only the meta-data and a pointer to the raw device, which could be treated as "thin qcow2".

The date_file and base image will be created together with a command "qemu-img create". 

Customers need a way to track meta-data of raw devices or files, for fetures like incremental backup. So add this function.

Since libvirt won't support external data file, we just cover installation and boot in qemu side. No other operation.

Reply:
As we discussed, added image deletion for data file. thanks.